src/coreclr/jit/valuenum.cpp

```diff
@@ -790,6 +790,29 @@ int ValueNumStore::EvalComparison<double>(VNFunc vnf, double v0, double v1)
                 break;
         }
     }
+    else // must be a VNF_ function
+    {
+        if (hasNanArg)
+        {
+            // unordered comparisons with NaNs always return true
+            return true;
+        }
+
+        switch (vnf)
+        {
+            case VNF_GT_UN:
+                return v0 > v1;
+            case VNF_GE_UN:
+                return v0 >= v1;
+            case VNF_LT_UN:
+                return v0 < v1;
+            case VNF_LE_UN:
+                return v0 <= v1;
+            default:
+                // For any other value of 'vnf', we will assert below
+                break;
+        }
+    }
```

Comment:
Do we have tests for comparisons with NaN (as a constant, not Xunit arg)?

Reply:
`ls *.cs -r | sls "(<=|==|>=|<|>|!=) ?(double|float).NaN ?(<=|==|>=|<|>|!=)"` in `src/tests` finds nothing, as does manual looking at `ls *.il -r | sls "ldc.r(4|8)"` which suggests we do not have great coverage for this. There are tests in libs that cover some `NaN` folding, but I would be very surprised to see them ever hit this path (unordered compares are generally seen when Jit inverts the condition, and those tests are too simple for that).

It is somewhat surprising that's the case in general, I will look into fixing this. Testing this particular change is tricky because it runs late'ish, hopefully I will be able to figure out something out though.